cpp/arcticdb/storage/azure/azure_mock_client.cpp

```diff
@@ -0,0 +1,137 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <azure/core/http/curl_transport.hpp>
+#include <azure/core.hpp>
+#include <azure/storage/blobs.hpp>
+
+#include <arcticdb/storage/storage_utils.hpp>
+#include <arcticdb/storage/azure/azure_client_wrapper.hpp>
+#include <arcticdb/storage/azure/azure_mock_client.hpp>
+#include <arcticdb/storage/object_store_utils.hpp>
+
+namespace arcticdb::storage::azure {
+
+std::string operation_to_string(AzureOperation operation){
+    switch (operation) {
+        case AzureOperation::READ: return "Read";
+        case AzureOperation::WRITE: return "Write";
+        case AzureOperation::DELETE: return "Delete";
+        case AzureOperation::LIST: return "List";
+        case AzureOperation::EXISTS: return "Exists";
+    }
+    util::raise_rte("Invalid Azure operation");
+}
+
+std::string MockAzureClient::get_failure_trigger(
+        const std::string& blob_name,
+        AzureOperation operation_to_fail,
+        Azure::Core::Http::HttpStatusCode error_to_fail_with) {
+    return fmt::format("{}#Failure_{}_{}", blob_name, operation_to_string(operation_to_fail), (int)error_to_fail_with);
+}
+
+Azure::Core::RequestFailedException get_exception(const std::string &message, Azure::Core::Http::HttpStatusCode status_code) {
+    auto rawResponse = std::make_unique<Azure::Core::Http::RawResponse>(0, 0, status_code, message);
+    auto exception = Azure::Core::RequestFailedException(rawResponse);
+
+    return exception;
+}
+
+std::optional<Azure::Core::RequestFailedException> has_failure_trigger(const std::string& blob_name, AzureOperation operation) {
+    auto failure_string_for_operation = "#Failure_" + operation_to_string(operation) + "_";
+    auto position = blob_name.rfind(failure_string_for_operation);
+    if (position == std::string::npos) return std::nullopt;
+
+    try {
+        auto failure_code_string = blob_name.substr(position + failure_string_for_operation.size());
+        auto status_code = Azure::Core::Http::HttpStatusCode(std::stoi(failure_code_string));
+        auto error_message = fmt::format("Simulated Error, message: #{}_{}", failure_string_for_operation, (int) status_code);
+
+        return get_exception(error_message, status_code);
+    } catch (std::exception& e) {
+        return std::nullopt;
+    }
+}
+
+void MockAzureClient::write_blob(
+        const std::string &blob_name,
+        arcticdb::Segment &&segment,
+        const Azure::Storage::Blobs::UploadBlockBlobFromOptions&,
+        unsigned int) {
+
+    auto maybe_exception = has_failure_trigger(blob_name, AzureOperation::WRITE);
+    if (maybe_exception.has_value()) {
+        throw maybe_exception.value();
```

Comment:
Why are you throwing this rather than the variant type idiom like `S3Result` in Ivo's mock?

Reply:
Because if an operation fails, Azure throws an exception whereas S3 returns error in `S3Result` object. For any of the `read`, `write` or other operation failures, Azure throws [RequestFailedException](https://azuresdkdocs.blob.core.windows.net/$web/cpp/azure-core/1.10.2/class_azure_1_1_core_1_1_request_failed_exception.html#:~:text=A%20RequestFailedException%20is%20sometimes). You can also look at the SDK code for [blob_client.UploadFrom()](https://github.com/Azure/azure-sdk-for-cpp/blob/1d09c535b249ccb139b4b37c3c4f4acd0b9211c2/sdk/storage/azure-storage-blobs/src/block_blob_client.cpp#L162) and [blob_client.DownloadTo()](https://github.com/Azure/azure-sdk-for-cpp/blob/1d09c535b249ccb139b4b37c3c4f4acd0b9211c2/sdk/storage/azure-storage-blobs/src/blob_client.cpp#L349) which are called by `RealAzureClient`.